Android/app/src/main/java/app/intra/ui/settings/ServerApprovalDialogFragment.java

```diff
@@ -0,0 +1,159 @@
+ * ordinary URL intent would not work.  The intent would open in a web browser, but this dialog is
+ * used when the VPN is active and failing.  The web browser would therefore fail to open the
+ * server's page.  Opening the page in a WebView here works, because Intra itself is whitelisted
+ * to bypass the VPN.
+ */
+public class ServerApprovalDialogFragment extends DialogFragment {
+  private static final String INDEX_KEY = "index";
+  private static final String SHOW_WEBSITE_KEY = "showWebsite";
+
+  /**
+   * Default constructor.  This is present because it's required by the framework, to handle
+   * re-creation of the dialog (e.g. on rotation).  Arguments are persisted in the arguments bundle.
+   */
+  public ServerApprovalDialogFragment() {
+    super();
+  }
+
+  /**
+   * This is the constructor used by MainActivity.  It creates a version of the dialog with three
+   * buttons: Accept, Cancel, and Visit Website.
+   * @param index The index of the recommended server in the built-in server database.
+   */
+  public ServerApprovalDialogFragment(int index) {
+    super();
+    setIndex(index);
+  }
+
+  /**
+   * Private constructor, used to show a version of the dialog that contains the website, with only
+   * two buttons (Accept and Cancel).
+   * @param index The index of the recommended server in the built-in server database.
+   * @param showWebsite If true, show the server's website in the body of the dialog.
+   */
+  private ServerApprovalDialogFragment(int index, boolean showWebsite) {
+    super();
+    setIndex(index);
+    setShowWebsite(showWebsite);
+  }
+
+  private Bundle getArgs() {
+    Bundle args = getArguments();
+    if (args != null) {
+      return args;
+    }
+    return new Bundle();
+  }
+
+  private void setIndex(int index) {
+    Bundle args = getArgs();
+    args.putInt(INDEX_KEY, index);
+    setArguments(args);
+  }
+
+  private int getIndex() {
+    return getArgs().getInt(INDEX_KEY);
+  }
+
+  private void setShowWebsite(boolean showWebsite) {
+    Bundle args = getArgs();
+    args.putBoolean(SHOW_WEBSITE_KEY, showWebsite);
+    setArguments(args);
+  }
+
+  private boolean getShowWebsite() {
+    return getArgs().getBoolean(SHOW_WEBSITE_KEY);
+  }
+
+  @Override
+  public @NonNull Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {
+    AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
+    LayoutInflater inflater = requireActivity().getLayoutInflater();
+
+    final int index = getIndex();
+    final String name = getResources().getStringArray(R.array.names)[index];
+    final String template = getResources().getString(R.string.found_fastest);
+    final String message = String.format(template, name);
+
+    final String link = getResources().getStringArray(R.array.server_websites)[index];
+    final boolean showWebsite = getShowWebsite();
+    final View body;
+    if (showWebsite) {
```

Comment:
Consider decomposing this method to be more readable, particularly the webview and layout creation could be abstracted away.

Reply:
Done